### PR TITLE
Catalog: local mode

### PR DESCRIPTION
**Pull request by nl0**
This PR contains `quilt_lite` backend implementation and related catalog changes (to be factored out).
Brief outline of the changes:

- [x] `quilt_lite`:
  - [x] package set-up
    - [x] initial set-up (`project.toml` with dependencies)
    - [x] rename to `quilt3_lite` (or `quilt3_local`?)
    - [x] ~~set up black (code formatting)~~ -- punt
    - [x] <del>poetry script to run the server</del> (punted)
    - [x] set up build and packaging infra
    - [x] fill package metadata
      - [x] [description](https://python-poetry.org/docs/pyproject/#description)
      - [x] [license](https://python-poetry.org/docs/pyproject/#license)
      - [x] [readme?](https://python-poetry.org/docs/pyproject/#readme)
      - [x] [homepage](https://python-poetry.org/docs/pyproject/#homepage)
      - [x] [repository](https://python-poetry.org/docs/pyproject/#repository)
      - [x] [documentation](https://python-poetry.org/docs/pyproject/#documentation)
      - [ ] [extra URLs?](https://python-poetry.org/docs/pyproject/#urls)
  - [x] add instructions on how to set up and run stuff to the README
  - [x] quilt registry implementation
    - [x] serve catalog files
    - [x] serve `config.json`
    - [x] proxy already-running catalog
    - [x] serve graphql
      - [x] stub
      - [x] packages: draft implementation
      - [x] packages: implement leftover resolvers
      - [x] optimize package queries
  - [x] python 3.7 compatibility
  - [x] run lambdas locally (#2550 #2549)

- [x] integrate with `quilt3`

- [x] catalog
  - [x] move package listing and browsing requests to graphql: #2552
  - [x] local-mode specific UI changes
    - [x] disable package manipulation UI
    - [x] disable Queries tab

### Factored-out PRs

- #2500 
- #2501 
- #2502 
- #2503 
- #2504 
- #2505 
- #2506 
- #2507 
- #2508 
- #2552

### CODE REVIEW

**Review by kevinemoore — COMMENTED** [review]
Quick notes:

- I was able to get `quilt_lite` running
- `npm run build` failed. I separately ran `npm install jsoneditor`. After that, the npm build worked
- several python dependencies were missing after poetry install. I installed them individually with pip: importlib_resources, aiofiles, ariadne, fastapi, and uvicorn
- Local mode should (I think) hide the `QUERIES` tab

**Comment by nl0** [comment]
@kevinemoore 

> `npm run build` failed. I separately ran `npm install jsoneditor`. After that, the npm build worked

probably you have an older `npm` version (which handles peer dependencies differently) -- consider upgrading to version 7 or 8

> several python dependencies were missing after poetry install. I installed them individually with pip: 
> importlib_resources, aiofiles, ariadne, fastapi, and uvicorn

that's very strange, because these are explicitly listed in `project.toml`, maybe it has something to do with incompatibility of python or poetry versions 🤷 

> Local mode should (I think) hide the `QUERIES` tab

yeah, this is planned, as well as hiding package creation UI and some other cloud-only features

File: quilt_lite/quilt_lite/graphql.py [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+    return base + "/" + child
+
+class PackageDirWrapper:
+    def __init__(self, path: str, subpkg: quilt3.Package):
+        self.path = path
+        self._subpkg = subpkg
+
+    @functools.cached_property
+    def metadata(self):
+        return self._subpkg._meta
```
Comment (sir-sigurd) [inline quilt_lite/quilt_lite/graphql.py:46]: This looks to me like overusing of `cached_property` since call of this function is quite cheap.

Consider
```suggestion
    @property
    def metadata(self):
        return self._subpkg._meta
```
or
```suggestion
    metadata = property(operator.attrgetter('_subpkg._meta'))
```

File: quilt_lite/pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,36 @@
+[tool.poetry]
+name = "quilt_lite"
+version = "0.1.0"
+description = ""
+authors = ["Alexei Mochalov <alexei@quiltdata.io>"]
+
+[tool.poetry.dependencies]
+# TODO: make it working under 3.7
+python = "^3.8"
+ariadne = {git = "https://github.com/mirumee/ariadne.git"}
```
Comment (sir-sigurd) [inline quilt_lite/pyproject.toml:10]: Any reason we don't use released version here?

File: quilt_lite/quilt_lite/graphql.py [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+
+
+class PackageListWrapper:
+    def __init__(self, bucket: str, filter: T.Optional[str] = None):
+        self.bucket = bucket
+        self.filter = filter
+
+    @functools.cached_property
+    def _package_list(self):
+        return quilt3.list_packages(f"s3://{self.bucket}")
```
Comment (sir-sigurd) [inline quilt_lite/quilt_lite/graphql.py:218]: It looks like a bad idea to cache iterator, since it's mutable and you can't get results from it twice.

File: quilt_lite/quilt_lite/graphql.py [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+    #             ]),
+    #           ),
+    #           R.join(''),
+    def _filter(self, package_name):
+        if not self.filter: return True
+        return self.filter in package_name
+
+    @functools.cached_property
+    def _filtered_package_list(self):
+        return list(filter(self._filter, self._package_list))
```
Comment (sir-sigurd) [inline quilt_lite/quilt_lite/graphql.py:239]: Consider this:

```suggestion
    @functools.cached_property
    def _filtered_package_list(self):
        packages = self._package_list
        if self.filter:
            packages = filter(lambda package_name: self.filter in package_name, packages)
        return list(packages)
```

File: quilt_lite/pyproject.toml (hunk 2 above)
Comment (sir-sigurd) [reply quilt_lite/pyproject.toml:10]: ping

File: quilt_lite/pyproject.toml [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+boto3 = "^1.10.0"
+quilt3 = {path = "../api/python/", develop = true}
```
Comment (sir-sigurd) [inline quilt_lite/pyproject.toml:12]:
```suggestion
quilt3 = "^3.6.0"
```

File: quilt_lite/pyproject.toml (hunk 2 above)
Comment (nl0) [reply quilt_lite/pyproject.toml:10]: yep, there was some reason, but i don't remember exactly, there were some dependency version conflicts or smth like this. i see the next version has been released since then, so it's probably ok to use it (0.14)

File: quilt_lite/pyproject.toml (hunk 5 above)
Comment (nl0) [reply quilt_lite/pyproject.toml:12]: feel free to commit this

File: quilt_lite/quilt_lite/graphql.py [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+    @functools.cached_property
+    def metadata(self):
+        return self._browse._meta
+
+    @functools.cached_property
+    def message(self):
+        return self.metadata.get("message")
+
+    @functools.cached_property
+    def totalEntries(self, *_):
```
[the remark this answers is not in the document]
Comment (nl0) [reply quilt_lite/quilt_lite/graphql.py:103]: no if it's a property

File: quilt_lite/quilt_lite/lambdas.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
+from base64 import b64decode, b64encode
+import importlib
+import sys
+
+from fastapi import FastAPI, HTTPException, Request, Response
+
+sys.path.extend([
+    '../lambdas/shared/',
+    '../lambdas/',
+])
```
Comment (sir-sigurd) [inline quilt_lite/quilt_lite/lambdas.py:10]: I think we should make it work without modifying `sys.path`.

Comment (nl0) [reply quilt_lite/quilt_lite/lambdas.py:10]: we definitely should. any ideas how?

Comment (sir-sigurd) [reply quilt_lite/quilt_lite/lambdas.py:10]: Symlink?

File: quilt3_local/quilt3_local/__init__.py [hunk 8]
```diff
@@ -0,0 +1 @@
+__version__ = '0.1.0'
```
Comment (sir-sigurd) [inline quilt3_local/quilt3_local/__init__.py:1]: I think this is not considered a good practice anymore, see https://www.python.org/dev/peps/pep-0396/#pep-rejection,

File: quilt3_local/pyproject.toml [hunk 9]
```diff
@@ -0,0 +1,39 @@
+[tool.poetry]
+name = "quilt3_local"
+version = "1.0.0-beta.11"
+description = ""
+authors = ["Alexei Mochalov <alexei@quiltdata.io>"]
```
Comment (sir-sigurd) [inline quilt3_local/pyproject.toml:5]: I think we should use the same as here:
https://github.com/quiltdata/quilt/blob/12fab1fb5e79b9cdfaddca1d31a76bbe47778df7/api/python/setup.py#L55-L56

File: quilt3_local/quilt3_local/api.py [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+api = fastapi.FastAPI()
+
+
+@api.get("/api/auth/get_credentials")
+def get_credentials():
+    """
+    Obtains credentials corresponding to your role.
+
+    Returns a JSON object with three keys:
```
Comment (sir-sigurd) [inline quilt3_local/quilt3_local/api.py:31]:
```suggestion
    Returns a JSON object with four keys:
```
or better don't specify count at all.

File: quilt3_local/quilt3_local/api.py [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+session_cred = boto3._get_default_session().get_credentials()
+
+
+api = fastapi.FastAPI()
+
+
+@api.get("/api/auth/get_credentials")
+def get_credentials():
+    """
+    Obtains credentials corresponding to your role.
```
Comment (sir-sigurd) [inline quilt3_local/quilt3_local/api.py:29]:
```suggestion
    Obtain credentials corresponding to your role.
```
per [PEP 257](https://www.python.org/dev/peps/pep-0257/):

> The docstring is a phrase ending in a period. It prescribes the function or method's effect as a command ("Do this", "Return that"), not as a description; e.g. don't write "Returns the pathname ...".

File: quilt3_local/README.md [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+resolution doesn't work or takes too long -- it contains cached dependency list
+(and other metadata) of the `quilt3` package which is being installed as a local
+dependency).
+
+### Catalog (node.js) environment set-up
+
+Catalog supports `node` version `16` and `npm` versions `7` and `8`.
+
+To install the dependencies simply run `npm i` from the catalog root directory
+(`../catalog/`).
```
Comment (sir-sigurd) [inline quilt3_local/README.md:37]: I feel like this should be in README for catalog and we should have just a reference here.

File: quilt3_local/README.md [hunk 13]
```diff
@@ -0,0 +1,68 @@
+# Quilt Lite
+
+Open source implementation of the Quilt registry that works in the local
+environment (not requiring AWS cloud services).
+
+## TL;DR
```
Comment (sir-sigurd) [inline quilt3_local/README.md:6]:
```suggestion
# Developer
## TL;DR
```

File: quilt3_local/quilt3_local/api.py [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+
+    Returns a JSON object with three keys:
+        AccessKeyId(string): access key ID
+        SecretAccessKey(string): secret key
+        SessionToken(string): session token
+        Expiration(ISO date string)
+    """
+    try:
+        if session_cred.token:
```
Comment (sir-sigurd) [inline quilt3_local/quilt3_local/api.py:38]: Would be cool if we could make local catalog work (at least partially) without any AWS credentials.

If it's not doable in this PR, we should clarify in README that AWS credentials are still needed. Probably we also should handle absent credentials case more gracefully, currently it crashes with
```
    if session_cred.token:
AttributeError: 'NoneType' object has no attribute 'token'
```

File: quilt3_local/quilt3_local/api.py [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+            return {
+                "AccessKeyId": session_cred.access_key,
+                "SecretAccessKey": session_cred.secret_key,
+                "SessionToken": session_cred.token,
+                "Expiration": getattr(session_cred, "expiry_time", None),
+            }
+        return sts_client.get_session_token()["Credentials"]
+    except ClientError as ex:
+        print(ex)
```
Comment (sir-sigurd) [inline quilt3_local/quilt3_local/api.py:47]: [`logging`](https://docs.python.org/3/library/logging.html) should be used here instead.

File: quilt3_local/pyproject.toml [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+include = ["quilt3_local/catalog_bundle/*"]
+
+[tool.poetry.dependencies]
+python = ">=3.7.1,<3.11"
+ariadne = "^0.14.0"
+boto3 = "^1.10.0"
+fastapi = "^0.70.0"
+uvicorn = {extras = ["standard"], version = "^0.15.0"}
```
Comment (sir-sigurd) [inline quilt3_local/pyproject.toml:13]: I think this should be in dev deps.

File: quilt3_local/quilt3_local/api.py (hunk 14 above)
Comment (nl0) [reply quilt3_local/quilt3_local/api.py:35]: > Would be cool if we could make local catalog work (at least partially) without any AWS credentials.

not sure that would work

> Probably we also should handle absent credentials case more gracefully, currently it crashes with

done

> If it's not doable in this PR, we should clarify in README that AWS credentials are still needed

will do

File: quilt3_local/README.md (hunk 12 above)
Comment (nl0) [reply quilt3_local/README.md:37]: yeah it makes sense

File: quilt3_local/README.md [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+
+Refer to the [catalog documentation](../catalog/).
+
+### Running
+
+You can either serve a static catalog bundle (produced by `npm run build`) or
+proxy static files from a running catalog instance.
+
+**NOTE**: you need valid AWS credentials available, see
+[boto3 docs](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/credentials.html#configuring-credentials") for details.
```
Comment (sir-sigurd) [inline quilt3_local/README.md:43]:
```suggestion
[boto3 docs](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/credentials.html#configuring-credentials) for details.
```

File: quilt3_local/README.md [hunk 18]
```diff
@@ -0,0 +1,80 @@
+# Quilt3 catalog: Local development mode
+
+Open source implementation of the Quilt3 registry that works in the local
+environment (not requiring AWS cloud services).
```
Comment (sir-sigurd) [inline quilt3_local/README.md:4]: Might make sense to note that S3 Select is required, since previously I think we used endpoint from open.

**Review by sir-sigurd — APPROVED** [review]

File: quilt3_local/README.md [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+files from the catalog running at `http://localhost:3001`, catalog URL
+configurable via `QUILT_CATALOG_URL` env var).
+
+### Building and publishing
+
+1. Bump package version in `pyproject.toml`
+
+2. Make sure you set up [credentials for `poetry`](https://python-poetry.org/docs/repositories/#configuring-credentials)
+
+3. Build and publish the package: `poetry publish --build`
```
[the remark this answers is not in the document]
Comment (nl0) [reply quilt3_local/README.md:80]: yeah that would be nice

File: quilt3_local/README.md [hunk 20]
```diff
@@ -0,0 +1,80 @@
+# Quilt3 catalog: Local development mode
```
[the remark this answers is not in the document]
Comment (nl0) [reply quilt3_local/README.md:1]: idk it's per @akarve's suggestion (language he proposed for the landing page) -- probably to emphasize that it's intended for development vs production use
